odoo/tools/populate.py

```diff
@@ -0,0 +1,72 @@
+import random
+from odoo.tools import pycompat
+
+def Random(seed):
+    r = random.Random()
+    r.seed(seed, version=2)
+    return r
+
+
+def format_str(val, counter, _): # todo format
+    if isinstance(val, str) and '%' in val:
+        return val % counter
+    return val
+
+def chain_generators(fields_generators):
+    generator = root_generator()
+    for (fname, field_generator) in fields_generators:
+        generator = field_generator(generator, fname)
+    return generator
+
+def root_generator():
+    yield {}, False
+    while True:
+        yield {}, True
+
+
+def randomize(vals, weights=None, seed=False, formater=format_str):
+    def generator(iterator, field_name, counter_offset=0):
+        r = Random(seed or field_name)
+        for counter, (values, complete) in enumerate(iterator):
+            val = r.choices(vals, weights)[0]
+            values[field_name] = formater(val, counter+counter_offset, values)
+            yield values, complete
+    return generator
+
+
+def cartesian(vals, weights=None, seed=False, formater=format_str):
+    def generator(iterator, field_name):
+        counter = 0
+        for counter, (values, complete) in enumerate(iterator):
+            if complete:
+                break
+            for val in vals:
+                yield {**values, field_name: formater(val, counter, values)}, False
+        yield from randomize(vals, weights, seed, formater)(iterator, field_name, counter)
+    return generator
+
+
+def iterate(vals, weights=None, seed=False, formater=format_str):
+    def generator(iterator, field_name):
+        counter = 0
+        for counter, (val, (values, _)) in enumerate(pycompat.izip(vals, iterator)): # iteratable order is important, shortest first
+            yield {**values, field_name: formater(val, counter, values)}, False
+        yield from randomize(vals, weights, seed, formater)(iterator, field_name, counter)
+    return generator
+
+
+def set_value(val, formater=format_str):
+    def generator(iterator, field_name):
+        for counter, (values, complete) in enumerate(iterator):
+            values[field_name] = formater(val, counter, values)
+            yield values, complete
+    return generator
+
+
+def call(function, seed=None):
+    def generator(iterator, field_name):
+        r = Random(seed or field_name)
+        for counter, (values, complete) in enumerate(iterator):
+            for val in function(values=values, counter=counter, complete=complete, pseudo_random=r):
+                yield {**values, field_name:val}, complete
+    return generator
```

Comment:
```suggestion
def compute(function, seed=None):
    def generate(iterator, field_name):
        r = Random(seed or field_name)
        for counter, (values, complete) in enumerate(iterator):
            val = function(values=values, counter=counter, complete=complete, random=r)
            yield {**values, field_name: val}, complete
    return generate
```

Reply:
done.

I just kept pseudo_random for now for one simple reason: if someone copy paste some code but not the function definition, I'm afraid that we could have some king of 'random is not defined error': ok lets import random. What do you think?

Reply:
Then name it `pseudo_random_to_not_be_confused_with_module_random` :wink: 
You are overprotective, IMHO.